src/Host.cpp

```diff
@@ -791,12 +1260,18 @@ bool Host::installPackage(const QString& fileName, int module)
             // now that the packageName changed, redo relevant checks to make sure it's still valid
             if (module) {
                 if (mActiveModules.contains(packageName)) {
-                    uninstallPackage(packageName, 2);
+                    uninstallPackage(packageName, 2, pErrorMessage);
                 }
             } else {
                 if (mInstalledPackages.contains(packageName)) {
                     // cleanup and quit if already installed
                     removeDir(_dir.absolutePath(), _dir.absolutePath());
+                    if (pErrorMessage) {
+                        *pErrorMessage = QLatin1String("nothing to do, package already installed");
+                    } else {
+                        postMessage(tr("[ ALERT ] - Package \"%1\" already seems to be installed, nothing to do.")
```

Comment:
I don't think we should say anything here at all. Alerts freak people out and there'll be people who won't understand this message - which doesn't actually have any information, so they'll just get freaked out over nothing.

Reply:
ALERT are - to my mind - advising that _something odd_ but not necessary __bad__ is happening, in this case it is a response to the user doing something which is NOT having an effect when they might expect it to - and after all it is saying why it isn't doing the thing.

Reply:
That's what they are to you, but my experience with users says that I can definitely point to people who will not understand this and get confused and frustrated.

If it's not doing the thing, then just don't do anything. It'll be clear it's not doing the thing. Showing alerts will just add confusion for most people, save for the engineers and programmers who also think the same.

Reply:
Okay will/have removed the console message (and left a comment noting why) so only the lua command will get the other message/branch of the `if` (as an part a __nil+errormessage__ response).

Reply:
I see a problem with this!

```lua
local installed, error = installPackage()
if not installed then return end

-- code here can safely assume package is installed
```

Nice and simple - except that we can't have that! It would actually have to be:

```
local installed, error = installPackage()
if error ~= "package is already installed" then return end

-- code here can safely assume package is installed
```

A bit less readable and more convoluted. The fact that a package is already installed when you ask the package to be installed should not be an error. The condition you asked for exists - good! Doesn't matter how we got there. If you'd like to be _super_ thorough, you can return true + message if you'd like.

Reply:
Humm, yeah, asking to install a _package_ that is already installed needs to be okay - though that then leads to the tiny niggling detail in that you do not know for certain that the (file?)name that you wanted installed is actually the particular file that you expected and not another copy from a different file... which if you are hacking around is not entirely impossible! 🙁 

💡 I wonder - ought we to provide an explicit lua `(bool) checkPackage("packageName", "optional path/filename to expected source")` for verification purposes - I noted that the current File mangler on my system (Dolphin) offers to calculate checksums {MD5/SHA1/SHA256} for files and I see that there is a LuaRocks [MD5](https://luarocks.org/modules/tomasguisasola/md5/1.2-1) and at least one Public Domain [MD5 implimentation](http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5) so we could even offer to return/use something like an MD5 checksum to finger-print such things in the future?

Whilst poking around in this area have found one or two other _package_ manager things which I feel I have to tidy up but I'll try that as a separate fix to get in first.

Reply:
Let's see if it comes up in practice - if we feel the need for it, we can add it. I'd rather not add complication for all the possibilities only to find them unused in practice!

Reply:
... actually, it would be simpler than _I_ thought given that [`QCryptographicHash`](http://doc.qt.io/qt-5/qcryptographichash.html) has supported the MD5 type since before Qt 5 came about!  However I agree that is something for another day/PR.